File: 프로그래머스/level3/외벽점검.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
 
int solution(int n, vector<int> weak, vector<int> dist) {
    int answer = -1;
    sort(dist.begin(), dist.end());
    for (int i = 0; i < weak.size(); i++) {
        int tmp = weak[0] + n;
        for (int j = 1; j < weak.size(); j++) {
            weak[j - 1] = weak[j];
        }
        weak[weak.size() - 1] = tmp;
        do {
            int w = 0;
            int d = 0;
            for (d = 0; d < dist.size(); d++) {
                int fin = weak[w] + dist[d];
                while (fin >= weak[w]) {
                    w++;
                    if (w == weak.size()) {
                        break;
                    }
                }
                if (w == weak.size()) {
                    break;
                }
            }
            if (w == weak.size()) {
                if (answer == -1 || d + 1 < answer) {
                    answer = d + 1;
                }
            }
        } while (next_permutation(dist.begin(), dist.end()));
    }
    return answer;
}

File: 프로그래머스/level3/외벽점검.js
const getPermutations= function (arr, selectNumber) {
    const results = [];
    if (selectNumber === 1) return arr.map((value) => [value]);

    arr.forEach((fixed, index, origin) => {
        const rest = [...origin.slice(0, index), ...origin.slice(index+1)];
        const permutations = getPermutations(rest, selectNumber - 1);
        const attached = permutations.map((permutation) => [fixed, ...permutation]);
        results.push(...attached);
    });
    return results;
};

function solution(n, weak, Dist) {
    var answer = 100;
    Dist.sort((a, b)=>b-a);
    let per = getPermutations(Dist, Dist.length);
    for(let k=0; k<weak.length; k++){
        // console.log(weak);
        per.forEach(dist=>{
            let idx = 0;
            let cnt = 0;
            // console.log(dist);
            for(let i=0; i<dist.length; i++){
                let now = weak[idx];
                let next = weak[idx]+dist[i];
                if(next>=n){
                    while(weak[idx]>=now||weak[idx]<=next%n){
                        idx++;
                    }
                }
                else{
                    while(weak[idx]>=now&&weak[idx]<=next){
                        idx++;
                    }
                }
                cnt++;
                // console.log(now, next, idx, cnt);
                if(idx>=weak.length){
                    answer = Math.min(answer, cnt);
                    break;
                }
            }
        })
        weak.push(weak[0]);
        weak.shift();
    }
    if(answer === 100)  answer = -1;
    return answer;
}